diff --git a/type/definition.js b/type/definition.js
index c2f989d84a0d8e82b9576c01339a41177b1386dd..173602c3f2231f34e50f3bccef1c541fc8ca58d0 100644
--- a/type/definition.js
+++ b/type/definition.js
@@ -54,6 +54,7 @@ exports.isUnionType = isUnionType;
 exports.isWrappingType = isWrappingType;
 exports.resolveObjMapThunk = resolveObjMapThunk;
 exports.resolveReadonlyArrayThunk = resolveReadonlyArrayThunk;
+exports.defineFieldMap = defineFieldMap;
 
 var _devAssert = require('../jsutils/devAssert.js');
 
